distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso

```diff
@@ -919,24 +933,39 @@ type Table
              example_set =
                  table = Examples.inventory_table
                  double_inventory = table.at "total_stock" * 2
-                 table.set "total_stock" double_inventory
-                 table.set "total_stock_expr" "2 * [total_stock]"
-    set : Text -> Column | Vector | Text -> Problem_Behavior -> Table
-    set self name column on_problems=Report_Warning = case column of
-        _ : Text ->
-            expression = self.evaluate column
-            if expression.is_error.not then self.set name expression on_problems else
-                on_problems.handle_errors expression fallback=self
-        _ : Vector -> self.set name (Column.from_vector name column)
-        _ : Column ->
-            Table.Value (self.java_table.addOrReplaceColumn (column.rename name . java_column))
+                 table.set double_inventory new_name="total_stock"
+                 table.set "2 * [total_stock]" new_name="total_stock_expr"
+    set : Text -> Column | Text -> Problem_Behavior -> Table ! Existing_Column | Missing_Column
+    set self column new_name=Nothing set_mode=Set_Mode.Add_Or_Update on_problems=Report_Warning =
+        resolved = case column of
+            _ : Text -> self.compute column on_problems
+            _ : Column -> column
+        renamed = if new_name.is_nothing then resolved else resolved.rename new_name
+        to_add = case set_mode of
+            Set_Mode.Add_Or_Update -> True
+            Set_Mode.Add -> if self.java_table.getColumnByName renamed.name . is_nothing then True else
+                Error.throw (Existing_Column.Error renamed.name)
+            Set_Mode.Update -> if self.java_table.getColumnByName renamed.name . is_nothing . not then True else
+                Error.throw (Missing_Column.Error renamed.name)
+
+        if to_add then Table.Value (self.java_table.addOrReplaceColumn renamed.java_column) else to_add
+
+    ## Given an expression, create a derived column where each value is the
+       result of evaluating the expression for the row.
 
-    ## PRIVATE
-    evaluate : Text -> Column
-    evaluate self expression =
+       Arguments:
+       - expression: The expression to evaluate.
+       - on_problems: Specifies how to handle problems with expression
+         evaluation.
+    compute : Text -> Problem_Behavior -> Column
+    compute self expression on_problems=Report_Warning =
         get_column name = self.at name
         make_constant value = Column.from_vector_repeated (UUID.randomUUID.to_text) [value] self.row_count
-        Expression.evaluate expression get_column make_constant "Standard.Table.Data.Column" "Column" Column.var_args_functions
+        new_column = Expression.evaluate expression get_column make_constant "Standard.Table.Data.Column" "Column" Column.var_args_functions
+        if new_column.is_error then new_column else
+            warnings = Warning.get_all new_column
+            rename = new_column.rename expression
```

Comment:
I see that here we are not 'sanitizing' the column name. Shall we try to make the generated column names consistent? Maybe let's sanitize them on both sides then? Ideally using a shared function, probably from a shared module like `Table_Helpers`.

I remember we had some task to do at some point to ensure that names of columns generated by operations on columns (like `+` etc.) should be consistent between the two backends (currently they are not); so ideally these new operations should not introduce new inconsistencies.